Automated thresholding suggestions (#315)
datacarpentry/image-processing: episodes/07-thresholding.md

![](fig/maize-root-cluster-histogram.png){alt='Grayscale histogram of the maize root image'}

The histogram has a significant peak around 0.2 and then a broader "hill" around 0.6 followed by a 

> smaller peak near 1.0. Looking at the grayscale image, we can identify the lower peak at 0.2 as the

Author: ```suggestion
smaller peak near 1.0. Looking at the grayscale image, we can identify the peak at 0.2 as the
```

> smaller peak near 1.0. Looking at the grayscale image, we can identify the peak at 0.2 as the
> background and the broader peak as the foreground.

Author: ```suggestion
smaller peak near 1.0. Looking at the grayscale image, we can identify the peak at 0.2 with the
background and the broader peak with the foreground.
```

Thus, this image is a good candidate for thresholding with Otsu's method.
The mathematical details of how this works are complicated (see
[the scikit-image documentation](https://scikit-image.org/docs/dev/api/skimage.filters.html#threshold-otsu)
if you are interested),
but the outcome is that Otsu's method finds a threshold value between the two peaks of a grayscale
histogram which might correspond well to the foreground and background depending on the data and 
application.

:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: instructor

The histogram of the maize root image may prompt questions from learners about the interpretation 
of the peaks and the broader region around 0.6. The focus here is on the separation of background 
and foreground pixel values. We note that Otsu's method does not work well 
for the image with the shapes used earlier in this episode, as the foreground pixel values are more 

> distributed. These examples could be augmented by a discussion of unimodal, bimodal, and multimodal

Author: ```suggestion
distributed. These examples could be augmented with a discussion of unimodal, bimodal, and multimodal
```

> histograms. While these points can lead to fruitful discussions, the text in this episode attempts 

Author: ```suggestion
histograms. While these points can lead to fruitful considerations, the text in this episode attempts 
```

![](fig/maize-root-cluster-histogram.png){alt='Grayscale histogram of the maize root image'}

> The histogram has a significant peak around 0.2, and a second,

Reviewer: ```suggestion
The histogram has a significant peak around 0.2 and then a broader "hill" around 0.6 followed by a smaller peak near 1.0. Looking at the grayscale image, we can identify the lower peak at 0.2 as the background and the larger pixel values as the foreground, but it is not so obvious what the threshold should be.
```

Reviewer 2: I recommend removing "lower" from your explanation above ("the lower peak at 0.2), because "lower" could have two different meanings in this context. I think "the peak at 0.2" is fine.

Reviewer 3: Yes, "lower peak" is confusing here. I would further suggest:
```diff
-we can identify the lower peak at 0.2 as the background
+we can identify the peak at 0.2 with the background
```
and
```diff
-the larger pixel values as the foreground
+the larger pixel values with the foreground
```
because we're associating pixel values with image features (rather than saying they're identical).

> shallower peak near 0.6.

Reviewer: ```suggestion
```

Thus, this image is a good candidate for thresholding with Otsu's method.
The mathematical details of how this works are complicated (see
[the scikit-image documentation](https://scikit-image.org/docs/dev/api/skimage.filters.html#threshold-otsu)
if you are interested),

> but the outcome is that Otsu's method finds a threshold value between
> the two peaks of a grayscale histogram.

Reviewer: ```suggestion
but the outcome is that Otsu's method finds a threshold value that separates the peaks of a grayscale histogram that might correspond to background and foreground pixel values.
```

Reviewer 3: I think that "a value between the peaks" is correct and that "separating" applies to "the foreground from the background" or vice-versa. So, I would suggest something along these lines:

> Otsu's method finds a threshold value between the two peaks of a grayscale histogram: This threshold thus separates higher from lower pixel values, which might correspond to the foreground and background respectively.

Author: ```suggestion
but the outcome is that Otsu's method finds a threshold value between
the two peaks of a grayscale histogram.
```
but the outcome is that Otsu's method finds a threshold value between the two peaks of a grayscale histogram which might correspond well to the foreground and background depending the data and application.

Note there is a third peak very close to 1.0 which corresponds to the white label and disk on the 
left hand side of the image. It would be a good idea to remove these white pixels 
before calculating the Otsu threshold. We will discuss how to do this later in the episode but for 

> now we will continue without removing them.

Reviewer 2: While it is undoubtedly accurate, I am not sure I prefer this to the current version: as currently written, I believe the idea is for Instructors to naively proceed on the assumption that the second-largest peak is the stuff we are most interested in in the image: sort of a binary way of thinking about the image as composed of background and foreground which might match quite well the learners' misconceptions. Then later to help learners re-assess that notion and realise that further selection needs to take place to isolate only the pixels of interest.

I would favour a note added to the "Ignoring more of the image" challenge, stating that really it would have been better practice to remove these pixels too before calculating the Otsu threshold in the first place. Interested to hear what others think @datacarpentry/image-processing-curriculum-maintainers

Reviewer 3: I agree with @tobyhodges's comment: From an educational viewpoint, I think that we want to proceed 'naively' indeed; we don't want to 'anticipate' oddities, but instead 'react' to these afterwards (let learners see them and update their mental models then).

Author: Great - now in a instructor note as per @uschille suggestion to avoid overloading the learners

> Note there is a third peak very close to 1.0 which corresponds to the white label and disk on the 
> left hand side of the image. It would be a good idea to remove these white pixels 
> before calculating the Otsu threshold. We will discuss how to do this later in the episode but for 
> now we will continue without removing them.

Reviewer: ```suggestion
:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: instructor
The histogram of the maize root image may prompt questions from learners about the interpretation of the peaks and the broader region around 0.6. The focus here is on the separation of background and foreground pixel values. We note that Otsu's method does not work well for the image with the shapes used earlier in this episode, as the foreground pixel values are more distributed. These examples could be augmented by a discussion of unimodal, bimodal, and multimodal histograms. While these points can lead to fruitful discussions, the text in this episode attempts to reduce cognitive load and deliberately simplifies the discussion.
::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
```

Reviewer 3: :+1: maybe
```diff
-augmented by
+augmented with
```
and
```diff
-fruitful discussions
+fruitful considerations
```
(to avoid some repetition)

    binary_mask = blurred_image > t

    # determine root mass ratio
    root_pixels = np.count_nonzero(binary_mask)
    w = binary_mask.shape[1]
    h = binary_mask.shape[0]

>     density = root_pixels / (w * h)

Reviewer 2: ❤️

    # perform binary thresholding to mask the white label and circle
    binary_mask = blurred_image < 0.95
    
    # perform automatic thresholding using only the pixels with value True in the binary mask
    t = ski.filters.threshold_otsu(blurred_image[binary_mask])
    
    # update binary mask to identify pixels which are both less than 0.95 and greater than t

>     binary_mask = np.logical_and(binary_mask, blurred_image > t)

Author: ```suggestion
    binary_mask = binary_mask & (blurred_image > t)
```

    # perform binary thresholding to mask the white label and circle
    binary_mask = blurred_image < 0.95
    
    # perform automatic thresholding using only the pixels with value True in the binary mask
    t = ski.filters.threshold_otsu(blurred_image[binary_mask])
    
    # update binary mask to identify pixels which are both less than 0.95 and greater than t

>     binary_mask = np.logical_and(binary_mask, blurred_image > t)

Reviewer: I do like this solution, however, I am concerned that the introduction of `np.logical_and` may increase cognitive load. An alternative might be 
```binary_mask = (t < blurred_image) & (blurred_image < 0.95)```
but this would also require understanding of Boolean logic, which is only skimmed over in earlier episodes. Any thoughts from @datacarpentry/image-processing-curriculum-maintainers or @datacarpentry/curriculum-advisors-image?

Reviewer 2: Perhaps we can add a `spoiler` block with a brief explanation of AND in this context? I will propose something...

Reviewer 2: See [my suggestion](https://github.com/datacarpentry/image-processing/pull/315/files#r1472764124) below

Reviewer 3: I think it's not bad to introduce a bit of Boolean logic when using a binary mask... But, maybe, if we wanted to avoid it, we could use compose the two masks?
```py
mask_white = blurred_image < 0.95
mask_thresh = blurred_image > t
mask_total = mask_thresh[mask_white]
```
And use the 'total' mask for selection? So we're not introducing new syntax.

Author: I think combining masks in this way is something that learners will use a lot if they carry on doing image proccessing/analysis after the course so worth introducing. Personally I think @uschille solution is the easiest to understand of the three so Im going with that.

and labels were skewing our root mass ratios:

```output
data/trial-016.jpg,0.046250166223404256
data/trial-020.jpg,0.05886968085106383
data/trial-216.jpg,0.13712117686170214
data/trial-293.jpg,0.13190342420212767
```

> 

Reviewer 2: ```suggestion
:::::::::::::::::::::::::::::::::::::::::: spoiler

### What is `&` doing in the example above?

The `&` operator above means that we have defined a logical AND statement. This combines the two tests of pixel intensities in the blurred image such that both must be true for a pixel's position to be set to `True` in the resulting mask.

| Result of `t < blurred_image` | Result of `blurred_image < 0.95 | Resulting value in `binary_mask` |
|----------|---------|---------|
| False | True | False |
| True | False | False |
| True | True | True |
 
Knowing how to construct this kind of logical operation can be very helpful in image processing. The University of Minnesota Library's [guide to Boolean operators](https://libguides.umn.edu/BooleanOperators) is a good place to start if you want to learn more.

::::::::::::::::::::::::::::::::::::::::::::::::::
Replace "The `&` operator" above with "The `np.logical_and` function" according to whatever solution ends up being merged.
```

and labels were skewing our root mass ratios:

```output
data/trial-016.jpg,0.046250166223404256
data/trial-020.jpg,0.05886968085106383
data/trial-216.jpg,0.13712117686170214
data/trial-293.jpg,0.13190342420212767
```
:::::::::::::::::::::::::::::::::::::::::: spoiler

### What is `&` doing in the example above?

The `&` operator above means that we have defined a logical AND statement. This combines the two tests of pixel intensities in the blurred image such that both must be true for a pixel's position to be set to `True` in the resulting mask.

> | Result of `t < blurred_image` | Result of `blurred_image < 0.95 | Resulting value in `binary_mask` |

Reviewer 2: ```suggestion
| Result of `t < blurred_image` | Result of `blurred_image < 0.95` | Resulting value in `binary_mask` |
```

I missed this closing backtick in my earlier suggestion.
